src/servers/shared_memory_manager.h

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2019-2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <mutex>
+#include <unordered_map>
+#include "src/core/trtserver.h"
+#include "src/core/server_status.pb.h"
+
+#ifdef TRTIS_ENABLE_GRPC
+#include "src/core/grpc_service.grpc.pb.h"
+#endif  // TRTIS_ENABLE_GRPC
+
+#ifdef TRTIS_ENABLE_GRPC_V2
+#include "src/core/grpc_service_v2.grpc.pb.h"
+#endif  // TRTIS_ENABLE_GRPC_V2
+
+#ifdef TRTIS_ENABLE_GPU
+#include <cuda_runtime_api.h>
+#endif  // TRTIS_ENABLE_GPU
+
+namespace nvidia { namespace inferenceserver {
+
+class SharedMemoryManager {
+ public:
+  SharedMemoryManager() = default;
+  ~SharedMemoryManager();
+
+  /// Add a shared memory block representing shared memory in system
+  /// (CPU) memory to the manager. Return TRTSERVER_ERROR_ALREADY_EXISTS
+  /// if a shared memory block of the same name already exists in the manager.
+  /// \param name The name of the memory block.
+  /// \param shm_key The name of the posix shared memory object
+  /// containing the block of memory.
+  /// \param offset The offset within the shared memory object to the
+  /// start of the block.
+  /// \param byte_size The size, in bytes of the block.
+  /// \return a TRTSERVER_Error indicating success or failure.
+  TRTSERVER_Error* AddSystemSharedMemory(
```

Comment:
How about changing to register / unregister terminology? The GRPC service is defined that way.

Reply:
Done